File: src/test/java/com/team10/backend/domain/user/unit/UserServiceTest.java
package com.team10.backend.domain.user.unit;

import com.team10.backend.domain.image.service.ImageUploadService;
import com.team10.backend.domain.user.dto.ProfileImageUpdateRequest;
import com.team10.backend.domain.user.dto.SellerResponse;
import com.team10.backend.domain.user.dto.SellerUpdateRequest;
import com.team10.backend.domain.user.dto.UserResponse;
import com.team10.backend.domain.user.dto.UserUpdateRequest;
import com.team10.backend.domain.user.entity.User;
import com.team10.backend.domain.user.repository.UserRepository;
import com.team10.backend.domain.user.service.UserService;
import com.team10.backend.global.exception.BusinessException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

import java.time.LocalDateTime;
import java.util.Optional;

import static com.team10.backend.global.exception.ErrorCode.NOT_SELLER;
import static com.team10.backend.global.exception.ErrorCode.USER_NOT_FOUND;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class UserServiceTest {

    @Mock
    private UserRepository userRepository;

    @Mock
    private ImageUploadService imageUploadService;

    @InjectMocks
    private UserService userService;

    @Test
    @DisplayName("사용자 개인정보 조회 - 성공")
    void getUserProfile_success() {
        // given
        User user = UserTestFixture.createBuyer();
        ReflectionTestUtils.setField(user, "id", 1L);

        when(userRepository.findById(1L)).thenReturn(Optional.of(user));

        // when
        UserResponse response = userService.getUserProfile(1L);

        // then
        assertNotNull(response);
        assertEquals(user.getName(), response.name);
    }

    @Test
    @DisplayName("판매자 개인정보 조회 - 성공")
    void getSellerProfile_success() {
        // given
        User user = UserTestFixture.createSeller();
        ReflectionTestUtils.setField(user, "id", 1L);
        ReflectionTestUtils.setField(user, "createdAt", LocalDateTime.now());
        ReflectionTestUtils.setField(user, "updatedAt", LocalDateTime.now());

        when(userRepository.findById(1L)).thenReturn(Optional.of(user));

        // when
        SellerResponse response = userService.getSellerProfile(1L);

        // then
        assertNotNull(response);
        assertEquals(user.getName(), response.name);
    }

    @Test
    @DisplayName("판매자 정보 조회 - 실패 (판매자가 아닌 경우)")
    void getSellerProfile_fail_notSeller() {
        // given
        User user = UserTestFixture.createBuyer();

        when(userRepository.findById(1L)).thenReturn(Optional.of(user));

        // when & then
        BusinessException ex = assertThrows(BusinessException.class,
                () -> userService.getSellerProfile(1L));

        assertEquals(NOT_SELLER, ex.getErrorCode());
    }

    @Test
    @DisplayName("사용자 개인정보 수정 - 성공")
    void updateMyUserProfile_success() {
        // given
        User user = UserTestFixture.createBuyer();
        user.updateProfileImage("https://old-image.test/profile.jpg");

        UserUpdateRequest request = new UserUpdateRequest(
                "새로운닉네임",
                "010-9999-9999",
                "부산"
        );

        when(userRepository.findById(1L)).thenReturn(Optional.of(user));
        ReflectionTestUtils.setField(user, "id", 1L);

        // when
        UserResponse response = userService.updateMyUserProfile(1L, request);

        // then
        assertNotNull(response);
        assertEquals("새로운닉네임", response.nickname);
        assertEquals("010-9999-9999", response.phoneNumber);
        assertEquals("부산", response.address);
    }

    @Test
    @DisplayName("사용자 프로필 이미지 수정 - 성공")
    void updateMyUserProfileImage_success() {
        User user = UserTestFixture.createBuyer();
        user.updateProfileImage("https://old-image.test/profile.jpg");

        ProfileImageUpdateRequest request =
                new ProfileImageUpdateRequest("https://new-image.test/profile.jpg");

        when(userRepository.findById(1L)).thenReturn(Optional.of(user));
        ReflectionTestUtils.setField(user, "id", 1L);

        UserResponse response = userService.updateMyProfileImage(1L, request);

        assertEquals("https://new-image.test/profile.jpg", response.imageUrl);
        verify(imageUploadService).deleteIfManaged("https://old-image.test/profile.jpg");
    }

    @Test
    @DisplayName("사용자 프로필 이미지 삭제 - 성공")
    void deleteMyUserProfileImage_success() {
        User user = UserTestFixture.createBuyer();
        user.updateProfileImage("https://old-image.test/profile.jpg");

        when(userRepository.findById(1L)).thenReturn(Optional.of(user));
        ReflectionTestUtils.setField(user, "id", 1L);

        UserResponse response = userService.deleteMyProfileImage(1L);

        assertNull(response.imageUrl);
        verify(imageUploadService).deleteIfManaged("https://old-image.test/profile.jpg");
    }

    @Test
    @DisplayName("판매자 개인정보 수정 - 성공")
    void updateMySellerProfile_success() {
        // given
        User user = UserTestFixture.createSeller();
        ReflectionTestUtils.setField(user, "id", 1L);
        ReflectionTestUtils.setField(user, "createdAt", LocalDateTime.now());
        ReflectionTestUtils.setField(user, "updatedAt", LocalDateTime.now());

        SellerUpdateRequest request = new SellerUpdateRequest(
                "새로운판매자",
                "010-8888-8888",
                "대구",
                "새로운 인사말입니다.",
                "999-999-99999"
        );

        when(userRepository.findById(1L)).thenReturn(Optional.of(user));

        // when
        SellerResponse response = userService.updateMySellerProfile(1L, request);

        // then
        assertNotNull(response);
        assertEquals("새로운판매자", response.nickname);
        assertEquals("대구", response.address);
        assertEquals("새로운 인사말입니다.", response.bio);
        assertEquals("999-999-99999", response.businessNumber);
    }

    @Test
    @DisplayName("존재하지 않는 사용자인 경우")
    void getUserEntity_fail_notFound() {
        when(userRepository.findById(1L)).thenReturn(Optional.empty());

        BusinessException ex = assertThrows(BusinessException.class,
                () -> userService.getUserProfile(1L));

        assertEquals(USER_NOT_FOUND, ex.getErrorCode());
    }

}

File: src/test/kotlin/com/team10/backend/domain/user/unit/UserServiceTest.kt
package com.team10.backend.domain.user.unit

import com.team10.backend.domain.image.service.ImageUploadService
import com.team10.backend.domain.user.dto.ProfileImageUpdateRequest
import com.team10.backend.domain.user.dto.SellerUpdateRequest
import com.team10.backend.domain.user.dto.UserUpdateRequest
import com.team10.backend.domain.user.repository.UserRepository
import com.team10.backend.domain.user.service.UserService
import com.team10.backend.fixture.UserFixture
import com.team10.backend.global.exception.BusinessException
import com.team10.backend.global.exception.ErrorCode
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.springframework.test.util.ReflectionTestUtils
import java.time.LocalDateTime
import java.util.Optional

@ExtendWith(MockitoExtension::class)
class UserServiceTest {
    @Mock
    lateinit  var userRepository: UserRepository

    @Mock
    lateinit  var imageUploadService: ImageUploadService

    @InjectMocks
    lateinit  var userService: UserService

    @Test
    @DisplayName("사용자 개인정보 조회 - 성공")
    fun getUserProfile_success() {
        // given
        val user = UserFixture.create()
        ReflectionTestUtils.setField(user, "id", 1L)

        whenever(userRepository.findById(1L))
            .thenReturn(Optional.of(user))

        // when
        val response = userService.getUserProfile(1L)

        // then
        Assertions.assertNotNull(response)
        Assertions.assertEquals(user.name, response.name)
    }

    @Test
    @DisplayName("판매자 개인정보 조회 - 성공")
    fun getSellerProfile_success() {
        // given
        val user = UserFixture.createWithSellerInfo()
        ReflectionTestUtils.setField(user, "id", 1L)
        ReflectionTestUtils.setField(user, "createdAt", LocalDateTime.now())
        ReflectionTestUtils.setField(user, "updatedAt", LocalDateTime.now())

        whenever(userRepository.findById(1L))
            .thenReturn(Optional.of(user))

        // when
        val response = userService.getSellerProfile(1L)

        // then
        Assertions.assertNotNull(response)
        Assertions.assertEquals(user.name, response.name)
    }

    @Test
    @DisplayName("판매자 정보 조회 - 실패 (판매자가 아닌 경우)")
    fun getSellerProfile_fail_notSeller() {
        // given
        val user = UserFixture.create()

        whenever(userRepository.findById(1L))
            .thenReturn(Optional.of(user))

        // when & then
        val ex = assertThrows<BusinessException> {
            userService.getSellerProfile(1L)
        }

        Assertions.assertEquals(ErrorCode.NOT_SELLER, ex.errorCode)
    }

    @Test
    @DisplayName("사용자 개인정보 수정 - 성공")
    fun updateMyUserProfile_success() {
        // given
        val user = UserFixture.create()
        ReflectionTestUtils.setField(user, "id", 1L)

        user.updateProfileImage("https://old-image.test/profile.jpg")

        val request = UserUpdateRequest(
            "새로운닉네임",
            "010-9999-9999",
            "부산"
        )

        whenever(userRepository.findById(1L))
            .thenReturn(Optional.of(user))

        // when
        val response = userService.updateMyUserProfile(1L, request)

        // then
        Assertions.assertNotNull(response)
        Assertions.assertEquals("새로운닉네임", response.nickname)
        Assertions.assertEquals("010-9999-9999", response.phoneNumber)
        Assertions.assertEquals("부산", response.address)
    }

    @Test
    @DisplayName("사용자 프로필 이미지 수정 - 성공")
    fun updateMyUserProfileImage_success() {
        // given
        val user = UserFixture.create()
        ReflectionTestUtils.setField(user, "id", 1L)

        user.updateProfileImage("https://old-image.test/profile.jpg")

        val request =
            ProfileImageUpdateRequest("https://new-image.test/profile.jpg")

        whenever(userRepository.findById(1L))
            .thenReturn(Optional.of(user))

        // when
        val response = userService.updateMyProfileImage(1L, request)

        // then
        Assertions.assertEquals(
            "https://new-image.test/profile.jpg",
            response.imageUrl
        )
        verify(imageUploadService)
            .deleteIfManaged("https://old-image.test/profile.jpg")
    }

    @Test
    @DisplayName("사용자 프로필 이미지 삭제 - 성공")
    fun deleteMyUserProfileImage_success() {
        val user = UserFixture.create()
        ReflectionTestUtils.setField(user, "id", 1L)

        user.updateProfileImage("https://old-image.test/profile.jpg")

        whenever(userRepository.findById(1L))
            .thenReturn(Optional.of(user))

        val response = userService.deleteMyProfileImage(1L)

        Assertions.assertNull(response.imageUrl)
        verify(imageUploadService)
            .deleteIfManaged("https://old-image.test/profile.jpg")
    }

    @Test
    @DisplayName("판매자 개인정보 수정 - 성공")
    fun updateMySellerProfile_success() {
        // given
        val user = UserFixture.createWithSellerInfo()
        ReflectionTestUtils.setField(user, "id", 1L)
        ReflectionTestUtils.setField(user, "createdAt", LocalDateTime.now())
        ReflectionTestUtils.setField(user, "updatedAt", LocalDateTime.now())

        val request = SellerUpdateRequest(
            "새로운판매자",
            "010-8888-8888",
            "대구",
            "새로운 인사말입니다.",
            "999-999-99999"
        )

        whenever(userRepository.findById(1L))
            .thenReturn(Optional.of(user))

        // when
        val response = userService.updateMySellerProfile(1L, request)

        // then
        Assertions.assertNotNull(response)
        Assertions.assertEquals("새로운판매자", response.nickname)
        Assertions.assertEquals("대구", response.address)
        Assertions.assertEquals("새로운 인사말입니다.", response.bio)
        Assertions.assertEquals("999-999-99999", response.businessNumber)
    }

    @Test
    @DisplayName("존재하지 않는 사용자인 경우")
    fun getUserEntity_fail_notFound() {
        whenever(userRepository.findById(1L))
            .thenReturn(Optional.empty())

        val ex = assertThrows<BusinessException> {
            userService.getUserProfile(1L)
        }

        Assertions.assertEquals(ErrorCode.USER_NOT_FOUND, ex.errorCode)
    }
}
